### PR TITLE
Add driver for the Futaba TOSD-5711BB VFDisplay on Elonex Artisan/Sca…

### PR DESCRIPTION
**Pull request by ajw107**
…leo Media Centre PCs
futaba.so driver and LCDd.conf change for the Futaba TOSD-5711BB VFD

### CODE REVIEW

File: server/drivers/futaba.c [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+	if (p->old_framebuf == NULL) {
+		report(RPT_ERR, "%s: unable to create old_framebuffer", drvthis->name);
+		return -1;
+	}
+
+	/* write to the file */
+	futaba_start_driver(drvthis);
+	memset(p->framebuf, ' ', p->width * p->height);
+	memcpy(p->old_framebuf, p->framebuf,p->width * p->height);
+	report(RPT_DEBUG, "%s: init() done", drvthis->name);
```
Comment (haraldg) [inline server/drivers/futaba.c]: Use debug() here, so this gets optimized away on non-debug builds ...
The same is true at more places.

File: server/drivers/futaba.c [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+ * \param drvthis  Pointer to driver structure.
+ * \param on       New backlight status.
+ */
+MODULE_EXPORT void
+futaba_backlight (Driver *drvthis, int on)
+{
+	//PrivateData *p = drvthis->private_data;
+	debug(RPT_DEBUG, "Backlight %s command ignored as not supported", (on) ? "ON" : "OFF");
+}
+
```
Comment (haraldg) [inline server/drivers/futaba.c:706]: The last two functions are not mandatory. Maybe just leave them out completely.

File: server/drivers/futaba.c [hunk 3]
```diff
@@ -0,0 +1,619 @@
...
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <stdio.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <libusb-1.0/libusb.h>
```
Comment (haraldg) [inline server/drivers/futaba.c]: This line fails on travis build. Drop the libusb-1.0 part.

File: acinclude.m4 [hunk 4]
```diff
@@ -136,6 +136,10 @@ dnl				else
 			DRIVERS="$DRIVERS EyeboxOne${SO}"
 			actdrivers=["$actdrivers EyeboxOne"]
 			;;
+                futaba)
+                        DRIVERS="$DRIVERS futaba${SO}"
+                        actdrivers=["$actdrivers futaba"]
+                        ;;
```
Comment (haraldg) [inline acinclude.m4:151]: You should check if libusb is actually available. See how it is done for the picolcd driver for example.

**Review by haraldg — CHANGES_REQUESTED** [review]
This looks quite good. The only thing that is missing is the section describing this driver in the user guide (even if it will be quite trivial) and the LCDd man page. See http://lcdproc.sourceforge.net/docs/current-dev.html#documentation also about coding style. How do you want to be listed in the list of contributors?

**Comment by ajw107** [comment]
Thank you so much for going through everything.  I’ll definitely make the changes you suggest and see how to update the documentation as well.  I’ll issue another pull request when I’m done.
Alex

File: server/drivers/futaba.c [hunk 5]
```diff
@@ -167,7 +187,7 @@ MODULE_EXPORT char *symbol_prefix = "futaba_";
 			{
 				if( futabaSendReport(p->my_handle, &my_report) )
 				{
-					fprintf(stderr,"Failed in futabaSendString()\n");
+					debug(RPT_ERR,"[%s] Failed to send report", drvthis->name);
```
Comment (haraldg) [inline server/drivers/futaba.c]: Are you sure you want debug() here and not report()?

File: server/drivers/futaba.c [hunk 6]
```diff
@@ -307,17 +460,23 @@ futaba_init (Driver *drvthis)
 	/* Allocate and store private data */
 	p = (PrivateData *) calloc(1, sizeof(PrivateData));
 	if (p == NULL)
+	{
+		report(RPT_ERR, "%s: Private Data is NULL", drvthis->name);
```
Comment (haraldg) [inline server/drivers/futaba.c]: This message is somewhat misleading. We are simply out of memory. (Should never happen on a modern system due to overcommitting by the kernel.)

**Review by haraldg — COMMENTED** [review]
This is looking good, found only two minor nits.

Since libusb0.1 is obsolete, I wouldn't have put effort into supporting it for a new driver, but of course it's a nice thing to have. This prompts the question: Do you feel motivated to work on issue #13? - You appear to be quite competent for that kind of work... ;)

Do you need any help with documentation?

**Comment by ajw107** [comment]
Thanks for the pointers, very much appreciated.  I’ve almost finished the documentation (I think, famous last words) so should be able to do another push soon.
Alex

**Comment by ajw107** [comment]
Hi
Sorry to bother you, but I’m having problems doing the docbook documentation.  I have to admit to never using docbook before, and only limited experience with xml, could you suggest an editor I could use?  Preferably as wysiwyg as possible, he he.
Alex

**Comment by haraldg** [comment]
> Sorry to bother you, but I’m having problems doing the docbook documentation. I have to admit to never using docbook before, and only limited experience with xml, could you suggest an editor I could use? Preferably as wysiwyg as possible, he he.

Writing my driver for lcdproc was my first encounter with docbook too ... I just looked at documentation of other drivers to use as templates and used the dev-guide for guidance as which files need changing. I just used vim.

I suggest you push whatever you come up with and I'll comment on your code. That's probably the easiest way. Just write here in the PR what your are still unhappy with. When we are done we will squash all commits into one to have a clean and complete commit in the repository.

File: docs/lcdproc-user/drivers/futaba.docbook [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    </note>
+  </sect2>
+
+  <sect2 id="futaba-requirements">
+    <title>Requirements</title>
+
+    <para>The driver is based on the <ulink
+    url="http://libusb.info/"><filename>libusb</filename></ulink> USB library
+    (Versions 1.0 and 0.1), which should make it work with Linux, the
+    different BSB variants as well as Darwin/MacOS X.</para>
```
Comment (haraldg) [inline docs/lcdproc-user/drivers/futaba.docbook]: BSB = BSD?

File: docs/lcdproc-user/drivers/futaba.docbook [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+    <note>
+      <para>When using a <filename>libusb</filename> based driver like
+      <code>futaba</code>, <application>LCDd</application> needs to be started
+      as root.</para>
+    </note>
```
Comment (haraldg) [inline docs/lcdproc-user/drivers/futaba.docbook:72]: Will dropping priviledges after initialization (the usual LCDd way) work or has LCDd run as root all the time?

File: docs/lcdproc-user/drivers/futaba.docbook [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+        <varlistentry>
+          <term><property>Size</property> =
+          <parameter>WIDTHxHEIGHT</parameter></term>
+
+          <listitem>
+            <para>Should not be changed from the default of 7x1.</para>
+          </listitem>
+        </varlistentry>
+
```
Comment (haraldg) [inline docs/lcdproc-user/drivers/futaba.docbook]: Looks like the size parameters are actually hardcoded in the driver and such an entry in the configuration file won't have any effect anyway. Just drop this from the documentation and LCDd.conf

File: docs/lcdproc-user/drivers/futaba.docbook [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+  <sect2 id="futaba-status">
+    <title>futaba driver status</title>
+
+    <para>All icons/indicators function well. Currently the ':' (colon) or '.'
+    (dot) in between characters do NOT operate. In a future update I hope to
+    get these to work, with three settings: one to enable/disable flashing;
+    one to set the flash rate; and one to choose between ':' (colon) or '.'
+    (dot).</para>
+  </sect2>
```
Comment (haraldg) [inline docs/lcdproc-user/drivers/futaba.docbook:101]: I have a smilar situation in the olimex_1x9 driver. Maybe you can take some inspiration from there or maybe you come up with something better, for me to copy ... ;)

File: server/drivers/futaba.c [hunk 11]
```diff
@@ -78,21 +93,22 @@
 
 #define VENDOR_ID 0x0547
 #define PRODUCT_ID 0x7000
-futabaDriver_t g_futabaDriver;
+//futabaDriver_t g_futabaDriver;
```
Comment (haraldg) [inline server/drivers/futaba.c:97]: Is there any point in keeping this line?

File: server/drivers/futaba.c [hunk 12]
```diff
@@ -207,292 +220,319 @@ MODULE_EXPORT char *symbol_prefix = "futaba_";
...
+			}
+			if ( found_dev == 1){
+				break;
 			}
 		}
+		if ( found_dev == 1){
+			break;
+		}
 	}
-done:
```
Comment (haraldg) [inline server/drivers/futaba.c]: IMO the goto was fine. Anyway, it's your code, so you decide.

File: docs/lcdproc-user/drivers/futaba.docbook [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+
+          <listitem>
+            <para>Should not be changed from the default of 7x1.</para>
+          </listitem>
+        </varlistentry>
+
+        <varlistentry>
+          <term><property>TitleSpeed</property> =
+          <parameter><replaceable>TITLESPEED</replaceable></parameter></term>
+
```
Comment (haraldg) [inline docs/lcdproc-user/drivers/futaba.docbook]: This is a parameter of the server, not the driver and need not be documented here. Also drop it from LCDd.conf together with the other server parameters. I missed that in the previous reviews.

**Review by haraldg — COMMENTED** [review]
This is going to be the perfect driver, thanks a lot for all the effort you put into this.

You need to add your new docbook file to lcdproc-user.docbook and drivers.docbook to get it actually included. 

After fixing the LCDd.conf issue, that I missed previously, this is ready to merge as far as I'm concerned. You can either squash and force-push into this PR or push the last fixes directly and I'll do the squashing on github.

**Comment by ajw107** [comment]
Hi
Just about to submit the final commit (well until I add extra features).  Thank you so much for all of your help, especially over the Christmas, an insane time of year, so mad in fact I can’t remember making the last commit and was shocked when I saw your email, he he…

I’ve made the changes you suggested, I must apologise for the errors in the docbook files, not only did I forgot to add the other auxiliary files, but my copy and pasting from the picolcd driver documentation got me into trouble with a few things.  You are right, the driver does not need to be run as root permanently, and I also took out the stuff about usbfs, as this is obviously old info (do you want me to do the same for the picolcd drivers docs)?  I’ll also supmit updated docs when the linux patch and script.xbmc.lcdproc patches are accepted.

On the LCDd.conf I removed everything from the futaba section, but I’ve added a note to the documentation about the TitleSpeed, mainly because I remember being very confused when I used the driver and everything zoomed past so quick and it taking me ages to find the property (although that may just be me, he he).

Sorry for leaving in one of commented out lines of code, at work we’re not allowed to delete any lines, and I tend to get into the habit of it.  I’ve also left the goto code using a flag instead to control the program flow, blame my programming teach at uni who had a serious thing against them and drummed it into us all, he he.

Finally, I’ve updated the Changelog.md and CREDITS.md files, apologies if this is not my place to do so, just thought it may be easier for you.  Oh, and could you squash my code into one commit, I ask as I’ve not actually done that before, and I have this horrible image of me accidently reverting it all by mistake, he he.
Yours
Alex

**Comment by haraldg** [comment]
No need to be sorry or apologize for anything. This is good stuff and we all learn as we go and reviews almost always find ways to improve the code - which is why it is a good thing to have them. (Maybe I'm a bit biased from working on the linux kernel though.)

You seem to be reasonable and competent. We probably should give you commit access if you want to continue to work on lcdproc. However you will need to ask @willfe about this, I don't have access level to add new people, I think.

> I also took out the stuff about usbfs, as this is obviously old info (do you want me to do the same for the picolcd drivers docs)?

Of course, whenever you see an area of improvement creating a PR or opening an issue is welcome. Especially also for the documentation. It is so much more difficult writing documentation: Code you can always judge on whether it works.

An other area, where your expertise would help a lot, is issue #13 as I don't have any experience with libusb myself. (I have no idea if this is an easy or a difficult job. I will tag a release soon, to have ready before the upcoming debian freeze.)

> Sorry for leaving in one of commented out lines of code, at work we’re not allowed to delete any lines, and I tend to get into the habit of it.

I think this is what we have git for. Though indeed this line would not end up in git, if I squash commits.

> Finally, I’ve updated the Changelog.md and CREDITS.md files, apologies if this is not my place to do so, just thought it may be easier for you.

That's fine. However updating ChangeLog.md from PRs tends to cause merge conflicts when multiple PRs are under review at the same time.

> Oh, and could you squash my code into one commit, I ask as I’ve not actually done that before, and I have this horrible image of me accidently reverting it all by mistake, he he.

Sure, I will do this shortly. However a long as you keep a branch with your history of changes, such mishaps are easy to repair.

Thanks a lot for your work.

**Comment by ajw107** [comment]
Don’t worry about commit access, I’m happy doing it this way, if you are.  I’ll also have a look at the picolcd documentation and issue 13 tomorrow, if that is okay?  From a quick look at it, and just to clarify, you want me to look through the other drivers, and any that ONLY have support for LIBUSB0.1, to add support for LIBUSB1.0 (and LIBFTDI1 by the looks of it, no experience of that particular library, but I’ll give it a go)?
Alex

**Comment by haraldg** [comment]
> I’ll also have a look at the picolcd documentation and issue 13 tomorrow, if that is okay?

Great.

> From a quick look at it, and just to clarify, you want me to look through the other drivers, and any that ONLY have support for LIBUSB0.1, to add support for LIBUSB1.0

Yes. If it is a lot of work or needs testing on real HW, then maybe just compile a list of drivers first and look on the Mailinglist for testers, before putting more effort into it. But lets disucss the details in the proper issue tracker.

> (and LIBFTDI1 by the looks of it, no experience of that particular library, but I’ll give it a go)?

I don't think there are many drivers using this and some patches are floating around, that did only the libftdi change without the libusb change. Turned out that this doesn't work, but the patches should be a good start.

**Pull request closed without merging**